File: app/src/main/java/fmrsabino/moviesdb/ui/search/SearchAdapter.java
package fmrsabino.moviesdb.ui.search;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import fmrsabino.moviesdb.R;
import fmrsabino.moviesdb.data.model.search.Result;
import fmrsabino.moviesdb.ui.detail.DetailActivity;

public class SearchAdapter extends RecyclerView.Adapter<SearchAdapter.ViewHolder> {
    private Context context;
    private List<Result> results = new ArrayList<>();
    private String baseUrl;

    @Inject
    public SearchAdapter(Context context) {
        this.context = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.search_item, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Result result = results.get(position);
        holder.textView.setText(results.get(position).title());
        Picasso.with(context).load(baseUrl + result.posterPath())
                .fit()
                .centerCrop()
                .into(holder.imageView);
    }

    @Override
    public int getItemCount() {
        return results.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @BindView(R.id.search_item_poster) ImageView imageView;
        @BindView(R.id.search_item_value) TextView textView;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            Result result = results.get(getAdapterPosition());
            Intent intent = new Intent(context, DetailActivity.class);
            intent.putExtra(DetailActivity.MOVIE_ID_EXTRA, result.id());
            context.startActivity(intent);
        }
    }

    public void setPosterBaseUrl(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    public void setResults(List<Result> results) {
        this.results.clear();
        if (results != null) {
            this.results.addAll(results);
        }
        notifyDataSetChanged();
    }
}

File: app/src/main/java/fmrsabino/moviesdb/ui/search/SearchAdapter.kt
package fmrsabino.moviesdb.ui.search

import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import fmrsabino.moviesdb.R
import fmrsabino.moviesdb.data.model.search.Result
import fmrsabino.moviesdb.ui.detail.DetailActivity
import kotlinx.android.synthetic.main.search_item.view.*
import java.util.*
import javax.inject.Inject

class SearchAdapter
@Inject
constructor(private val context: Context) : RecyclerView.Adapter<SearchAdapter.ViewHolder>() {
    private val results = ArrayList<Result>()
    var baseUrl: String? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.search_item, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(results[position])

    override fun getItemCount() = results.size

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            val result = results[adapterPosition]
            val intent = Intent(context, DetailActivity::class.java)
            intent.putExtra(DetailActivity.MOVIE_ID_EXTRA, result.id)
            context.startActivity(intent)
        }

        fun bind(result: Result) {
            itemView.search_item_value.text = result.title
            Picasso.with(context)
                    .load(baseUrl + result.posterPath)
                    .fit()
                    .centerCrop()
                    .into(itemView.search_item_poster)
        }
    }

    fun setResults(results: List<Result>?) {
        this.results.clear()
        results?.let { this.results.addAll(it) }
        notifyDataSetChanged()
    }
}
